faunadb-scala/src/main/scala/faunadb/streaming/BodyValueFlowProcessor.scala

```diff
@@ -0,0 +1,88 @@
+package faunadb.streaming
+
+import java.nio.ByteBuffer
+import java.util
+import java.util.concurrent.{Flow, SubmissionPublisher}
+
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.ObjectMapper
+import faunadb.errors.PermissionDeniedException
+import faunadb.values.{StringV, VSuccess, Value}
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters.asScalaIteratorConverter
+import scala.util.Try
+
+private [faunadb] class BodyValueFlowProcessor(syncLastTxnTime: Long => Unit) extends SubmissionPublisher[Value] with Flow.Processor[util.List[ByteBuffer], Value] {
+  private val log = LoggerFactory.getLogger(getClass)
+  private val json = new ObjectMapper
+  private var subscription: Flow.Subscription = null
+  private var subscriber: Flow.Subscriber[_ >: Value] = null
+
+  // We do not request data from the publisher until we have one subscriber
+  override def subscribe(subscriber: Flow.Subscriber[_ >: Value]): Unit = {
+    if (this.subscriber == null) {
+      this.subscriber = subscriber
+      super.subscribe(subscriber)
+      requestOne()
+    } else
+      throw new IllegalStateException("BodyValueFlowProcessor can have only one subscriber")
+  }
+
+  override def onSubscribe(subscription: Flow.Subscription): Unit =
+    this.subscription = subscription
+
+  override def onNext(item: util.List[ByteBuffer]): Unit = {
+    import java.nio.charset.StandardCharsets
+    val text = item.iterator().asScala.map(StandardCharsets.US_ASCII.decode(_).toString).mkString("")
+    Try {
+      val jsonNode = json.readTree(text)
+      val value = json.treeToValue[Value](jsonNode, classOf[Value])
+
+      // syncLastTxnTime if possible
+      value("txnTS").to[Long].toOpt.foreach(syncLastTxnTime)
+
+      // handle error in stream
+      isUnrecoverableError(value) match {
+        case None => submit(value)
+        case Some(unrecoverableError) =>
+          subscriber.onError(unrecoverableError) // notify subscriber stream
+          subscription.cancel() // cancel subscription on the request body
+      }
+    }.recover {
+      case e: JsonProcessingException => log.error(s"could not parse JSON event $text", e)
+      case e: Throwable => log.error(s"could not handle event $text", e)
+    }
+
+    requestOne()
+  }
+
+  private def isUnrecoverableError(event: Value): Option[Throwable] = {
+    (event("event"), event("data", "code")) match {
+      case (VSuccess(StringV("error"), _), VSuccess(StringV("permission denied"), _)) =>
+        val description = event("data", "description").map(_.toString).getOrElse("no description available")
+        Some(PermissionDeniedException(None, s"code: permission denied, description:$description"))
```

Comment:
I was checking how the Exception messages are normally parsed when going through the `QueryError` flow, and it is as follows:

```
s"$code: $description"
```

From: 

https://github.com/fauna/faunadb-jvm/blob/1b1410f784a283664cb3c88396ce4d5ab06f6cba/faunadb-scala/src/main/scala/faunadb/errors/FaunaException.scala#L7

Can we change this to the following here?

```scala
case (VSuccess(StringV("error"), _), VSuccess(StringV(code), _)) if code == "permission denied" =>
  val description = event("data", "description").map(_.toString).getOrElse("no description available")
  Some(new PermissionDeniedException(s"$code: $description"))
```